File: projects/akka/tcpOSGI/src/main/scala/com/osgi/Activator.scala
package com.osgi

import org.osgi.framework.{BundleActivator, BundleContext, ServiceReference}
import java.io.{BufferedReader, IOException, InputStreamReader, PrintWriter}
import java.net.{ServerSocket, Socket}

import com.helpers.UserHelper
import com.liferay.portal.kernel.model.{Company, User}
import com.liferay.portal.kernel.service.{CompanyLocalServiceUtil, UserLocalService, UserLocalServiceUtil}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.collection.JavaConversions._
import scala.concurrent.Future
import akka.actor._
import akka.osgi.{ActorSystemActivator, OsgiActorSystemFactory}
import com.typesafe.config.ConfigFactory
import org.osgi.service.component.annotations.Reference
import org.osgi.util.tracker.ServiceTracker

import scala.concurrent.duration.Duration
import scala.concurrent.Promise

case class CreateUser(options:Map[String,Any])
case class DeleteUser(options:Map[String,Any])
case class ShowUser(options:Map[String,Any])
case class ListUsers()
case class HelpCommand()

class Activator  extends ActorSystemActivator with BundleActivator  with Parser {


	def configure(context: BundleContext, system: ActorSystem) {
		println(s"\n\nmy actor system name is called ${system.name} \n\n ")
    // optionally register the ActorSystem in the OSGi Service Registry
    //registerService(context,system)
    context.registerService(system.getClass.getName,system,null)
		println(s"Configuring Akka")
  }

	override def start(context: BundleContext){
    super.start(context)
		implicit val cont=context
		println("Starting server")
		Future{runServer}
	}

	override def stop(context: BundleContext): Unit ={
		println("Stopping server")
	}

	private val portNumber=4321
	private val keepRunning=true

	def usage=s"""
							 			|
							 			|Usage examples:
							 			|	${usageExamples.foldLeft("")((s,acc)=>s"$s$acc \r\t")}
							 			|	<there will be more implemented soon>
		""".stripMargin

	//TODO: add schedule jobs usage example??
	val usageExamples=List(
		"help",
		"user(s) create -username=<my_desired_username> -password=<my_password> -email=<my_email>",
		"create user -username=<my_desired_username> -password=<my_password> -email=<my_email>",
		"user delete -username=<my_desired_username>",
		"delete users -email=<my_email>",
		"users list",
		"list users # order doesn't matter here",
		"user count"
		//"download <http://mysite.com/myjar.jar>"
	)




	def runServer (implicit cont:BundleContext)={
    val references=cont.getAllServiceReferences("akka.actor.ActorSystemImpl",null)
    println(s"found = ${references.size} references to akka.actor.ActorSystemImpl")
		//val mySystem = OsgiActorSystemFactory(cont, getActorSystemConfiguration(cont)).createActorSystem("osgiActorSystem")
		val mySystem:ActorSystem=cont.getService(references(0)).asInstanceOf[ActorSystem]
    println(s"\n\n\n\nactor system name = ${mySystem.name}\n\n\n\n")

		val helloActor = mySystem.actorOf(Props[HelloActor], name = "helloActor")
		helloActor ! "hello"
		val serverSocket = new ServerSocket(portNumber)
		val clientSocket = serverSocket.accept()
		implicit val out = new PrintWriter(clientSocket.getOutputStream(), true)
		out.print("_________________________________\nWelcome to the FAKE gogo shell!!\n\n\nf! ")
		out.flush()
		val in = new BufferedReader( new InputStreamReader(clientSocket.getInputStream()))
		var line: String =in.readLine()
		val uservice: UserLocalService =UserLocalServiceUtil.getService
		while ( line != null && keepRunning) {
			if(line.nonEmpty){
				parse(pLiffey, line.mkString) match {
					case Success(matched,_) => {
						implicit val usersservice:UserLocalService=uservice
						matched match{
							case CreateUser(opts)=> UserHelper.createUser(opts)
							case DeleteUser(opts)=> UserHelper.deleteUser(opts)
							case ShowUser(opts)=> UserHelper.showUser(opts)
							case ListUsers=> UserHelper.listUsers
							case HelpCommand=> out.println(s"HELP COMMAND\n$usage")
							case _=> out.println( s"$matched UNIMPLEMENTED OPTION (please report the bug)" )
						}
					}
					case _=> out.println( s"UNRECOGNIZED COMMAND: $line" )
				}
			}
			out.print("f! ")
			out.flush()
			line=in.readLine()

		}
		serverSocket.close()
		clientSocket.close()
	}

	def findUser(email: String,out: PrintWriter): Unit={
		try{
			for(c<- CompanyLocalServiceUtil.getCompanies()){
				val u= UserLocalServiceUtil.getUserByEmailAddress( c.getCompanyId() , email)
				out.println(s"${u.getFirstName} - ${u.getCompanyId}")
			}
		}catch{ case e:Throwable => out.println("") }
	}

}

class HelloActor extends Actor {
  def receive = {
    case "hello world" =>
			sender ! "hello my friend"
			println("hello back at you")
    case _       => println("huh?")
  }
}

File: projects/client/src/com/examples/console/Activator.java
package com.examples.console;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Random;

import org.osgi.framework.BundleActivator;
import org.osgi.framework.BundleContext;
import org.osgi.framework.InvalidSyntaxException;
import org.osgi.framework.ServiceReference;

import com.liferay.training.service.hello.HelloService;

public class Activator implements BundleActivator {
	Random rand=new Random();
	private HelloService[] services;
	
	private int portNumber=4321;
	private boolean keepRunning=true;
	private BundleContext bundleContext;

	@Override
	public void start(BundleContext bc) throws Exception {
		bundleContext=bc;
		int c = setServices();
		for(int i=0;i<10;i++){System.out.println(services[rand.nextInt(c)].say("hello")); }
		startTCPServer(portNumber);
	}


	private void startTCPServer(int port) {
		System.out.println("starting tcp server on port "+port);
		new Thread(()->{try { runServer(port);
		} catch (Exception e) { 
			System.out.println("couldn't start TCP server");
			e.printStackTrace(); 
			}}).start();
	}


	private int setServices(int impl) throws InvalidSyntaxException {
		setServices();
		HelloService[] newServices = new HelloService[1];
		try{
			newServices[0]=services[impl];
			services=newServices;
			System.out.println("services set to service"+impl);
			return 1;
		}catch(Exception e){
			System.out.println("could not set services to service"+impl);
			return services.length;
		}
	}
	private int setServices() throws InvalidSyntaxException {
		ServiceReference<?>[] refs = bundleContext.getServiceReferences(HelloService.class.getName(), null);
		int c=0;
		HelloService[] newServices=new HelloService[refs.length];
		for(ServiceReference<?>r:refs)newServices[c++]=(HelloService)bundleContext.getService(r);
		services=newServices;
		System.out.println(c+" services set");
		return c;
	}


	public void stop(BundleContext context) throws Exception {
		System.out.println("stopping client...");
		//service = null;
	}

	public void runServer(int port) throws IOException {
		System.out.println("[Hello Client] running server on port "+port);
		ServerSocket serverSocket = null;
		try {
			serverSocket = new ServerSocket(port);
		  Socket clientSocket = serverSocket.accept();
		  PrintWriter out = new PrintWriter(clientSocket.getOutputStream(), true);
		  out.print("_________________________________\nWelcome to the FAKE gogo shell!!\n\n\nf! ");
		  out.flush();
	      System.out.println("socket created");
	      BufferedReader in = new BufferedReader( new InputStreamReader(clientSocket.getInputStream()));
	      String line=in.readLine().trim();
	      while ( line != null && keepRunning) {
	    	  System.out.println(String.format("read line of %s chars=%s", line.length(),line));
	    	  switch(line.trim().split(" ")[0]){
	    	  case "@": setServices(); break;
	    	  case "@1": setServices(0); break;
	    	  case "@2": setServices(1); break;
	    	  case "@port": 
	    		Integer newPort=null;
	    		try{
	    			newPort=Integer.parseInt(line.trim().split(" ")[1]);
	    			if(newPort!=null)startTCPServer(newPort);
	    		}catch(Exception e){
	    			System.out.println("couldn't start new TCP server when asked this: "+line);
	    		}
	    		break;
	    	  default:
	    		  if(line.length()>0) out.println(services[rand.nextInt(services.length)].say(line));
	    		  else out.println(services[rand.nextInt(services.length)].say());
	    		  break;
	    	  }
	    	  out.print("f! ");
	          out.flush();
	          line=in.readLine().trim();
	      }
	      serverSocket.close();
	      clientSocket.close();
		}catch(Exception e){
	      serverSocket.close();
		}
	}
	
	public static void main(String[] args){
		Activator activator=new Activator();
		activator.startTCPServer(1234);
		activator.startTCPServer(1235);
		activator.startTCPServer(1236);
	}


}